target/arc/semfunc-v3.c

```diff
@@ -14849,3 +14847,1279 @@ arc_gen_ATLDL(DisasCtxt *ctx, TCGv b, TCGv c)
+
+    tcg_gen_mov_tl(dest, new_dest);
+
+    tcg_temp_free(address);
+    tcg_temp_free(l_src1);
+    tcg_temp_free(l_src2);
+    tcg_temp_free(new_dest);
+
+    return ret;
+}
+
+/*
+ * FST32
+ *    Variables: @src1, @src2, @dest
+ *    Functions: arc_gen_set_memory
+ * --- code ---
+ * {
+ *   addr = shift_scale(src1, src2)
+ *   mem(addr) <= dest[0:32]
+ * }
+ */
+int
+arc_gen_FST32(DisasCtxt *ctx, TCGv src1, TCGv src2, TCGv dest)
+{
+    int ret = DISAS_NEXT;
+    int ZZ = 0; /* 32bit */
+    TCGv address = tcg_temp_local_new();
+    TCGv l_src1 = tcg_temp_local_new();
+    TCGv l_src2 = tcg_temp_local_new();
+
+    arc_gen_ldst_pre(ctx, address, src1, src2, ZZ);
+
+    tcg_gen_mov_tl(l_src1, src1);
+    tcg_gen_mov_tl(l_src2, src2);
+
+    setMemory(address, ZZ, dest);
+
+    arc_gen_ldst_post(ctx, src1, l_src1, l_src2);
+
+    tcg_temp_free(address);
+    tcg_temp_free(l_src1);
+    tcg_temp_free(l_src2);
+
+    return ret;
+}
+
+/*
+ * FLD64
+ *    Variables: @src1, @src2, @dest
+ *    Functions: arc_gen_get_memory
+ * --- code ---
+ * {
+ *   addr = shift_scale(src1, src2)
+ *   dest <= mem(addr)[0:64]
+ * }
+ */
+int
+arc_gen_FLD64(DisasCtxt *ctx, TCGv src1, TCGv src2, TCGv dest)
+{
+    int ret = DISAS_NEXT;
+    int ZZ = 3; /* 64 bit */
+    TCGv address = tcg_temp_local_new();
+    TCGv l_src1 = tcg_temp_local_new();
+    TCGv l_src2 = tcg_temp_local_new();
+    TCGv new_dest = tcg_temp_local_new();
+
+    arc_gen_ldst_pre(ctx, address, src1, src2, ZZ);
+
+    tcg_gen_mov_tl(l_src1, src1);
+    tcg_gen_mov_tl(l_src2, src2);
+
+    getMemory(new_dest, address, ZZ);
+
+    arc_gen_ldst_post(ctx, src1, l_src1, l_src2);
+
+    tcg_gen_mov_tl(dest, new_dest);
+
+    tcg_temp_free(address);
+    tcg_temp_free(l_src1);
+    tcg_temp_free(l_src2);
+    tcg_temp_free(new_dest);
+
+    return ret;
+}
+
+/*
+ * FST64
+ *    Variables: @src1, @src2, @dest
+ *    Functions: arc_gen_set_memory
+ * --- code ---
+ * {
+ *   addr = shift_scale(src1, src2)
+ *   mem(addr) <= dest[0:32]
+ * }
+ */
+int
+arc_gen_FST64(DisasCtxt *ctx, TCGv src1, TCGv src2, TCGv dest)
+{
+    int ret = DISAS_NEXT;
+    int ZZ = 3; /* 64 bit */
+    TCGv address = tcg_temp_local_new();
+    TCGv l_src1 = tcg_temp_local_new();
+    TCGv l_src2 = tcg_temp_local_new();
+
+    arc_gen_ldst_pre(ctx, address, src1, src2, ZZ);
+
+    tcg_gen_mov_tl(l_src1, src1);
+    tcg_gen_mov_tl(l_src2, src2);
+
+    setMemory(address, ZZ, dest);
+
+    arc_gen_ldst_post(ctx, src1, l_src1, l_src2);
+
+    tcg_temp_free(address);
+    tcg_temp_free(l_src1);
+    tcg_temp_free(l_src2);
+
+    return ret;
+}
+
+/*
+ * FLDD64
+ *    Variables: @src1, @src2, @dest
+ *    Functions: arc_gen_get_memory, nextFPURegWithNull
+ * --- code ---
+ * {
+ *   addr = shift_scale(src1, src2)
+ *   dest <= mem(addr)[0:64]
+ *   dest2 = nextFPUReg(dest)
+ *   dest2 <= mem(addr + 8)[0:64]
+ * }
+ */
+int
+arc_gen_FLDD64(DisasCtxt *ctx, TCGv src1, TCGv src2, TCGv dest)
```

Comment:
Can we use QEMU's 128 bit loads ?

Reply:
and what that would be?

Reply:
If possible, this can and should also be implemented in other instructions such as LDD/STD so I suggest leaving it for an eventual optimization round